### PR TITLE
MNT: Pin numpy<2 for v5.3.x

### PR DESCRIPTION
**Pull request by pllim**
### Description

I think it is reasonable to say that for v5.3.x, we will not support NumPy 2.0 or later? The support for v5.3.x will end anyway when v6.0.x is cut, which is schedule to be by the end of 2023.

Motivations:

* We have to keep playing catch up with NumPy 2.0 breaking changes in `main`. Backport is just more churn.
* Such capping is already being discussed in #15206 , so this basically implements @neutrinoceros's request for v5.3.x (v5.0.x is already capped) though that discussion is still open for upcoming v6.0.x.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/modeling/tests/test_constraints.py [hunk 1]
```diff
@@ -211,7 +210,7 @@ def test_bounds_gauss2d_lsq(self, fitter):
             ctx2 = nullcontext()
             if isinstance(fitter, fitting.TRFLSQFitter):
                 ctx = np.errstate(invalid="ignore", divide="ignore")
-                if not NUMPY_LT_2_0 or not SCIPY_LT_1_11_2:
+                if not SCIPY_LT_1_11_2:
```
Comment (pllim) [inline astropy/modeling/tests/test_constraints.py:213]: Hmm... Looks like this warning somehow went away again in dev scipy?

**Comment by mhvk** [comment]
I think this makes a lot of sense!

**Comment by pllim** [comment]
@mhvk , wanna approve? 😉

**Review by mhvk — APPROVED** [review]
@pllim - am happy to approve this, though perhaps good to have someone from the @astropy/astropy-project-release-team approve too? Though, in  a way, this really just makes sense...

File: pyproject.toml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 requires = ["setuptools",
             "setuptools_scm>=6.2",
             "cython==0.29.34",
-            "oldest-supported-numpy",
+            "numpy>=1.25,<2",
```
Comment (jdavies-st) [inline pyproject.toml:5]: Is there a reason to have the minimum build version higher than the minimum runtime version? That’s generally the thing that has caused problems in the past and is the reason for `oldest-supported-bumpy`.

Comment (neutrinoceros) [reply pyproject.toml:5]: Numpy 1.25 is the first version that supports backward compatible builds. The release notes explicitly mention that the mechanism makes oldest-supported-numpy not needed

**Comment by pllim** [comment]
@astrofrog gave verbal approval on Slack, so merging. Thanks, everyone!

**Comment by astrofrog** [comment]
What about v5.0.x? Should we make the same change there?

**Comment by pllim** [comment]
I already did that much earlier due to something else. So now both of them are pinned.

**Pull request merged**